src/rust-crypto/rust-crypto.ts

```diff
@@ -1357,6 +1357,50 @@ export class RustCrypto extends TypedEventEmitter<RustCryptoEvents, RustCryptoEv
         }
     }
 
+    /**
+     * Callback for `OlmMachine.registerReceiveSecretCallback`.
+     *
+     * Called by the rust-sdk whenever a secret is received.
+     * The gossipped secrets are received using the `m.secret.send` event type
+     * and are guaranteed to have been received over a 1-to-1 Olm
+     * Session from a verified device.
+     *
+     * The only secret this is currently broadcasted is the `m.megolm_backup.v1`.
+     *
+     * @param name - the secret name
+     * @param base64 - the secret value base 64 encoded
+     */
+    public async onReceiveSecret(name: string, value: string): Promise<void> {
+        this.logger.debug(`onReceiveSecret: Received secret ${name}`);
+        if (name === "m.megolm_backup.v1") {
+            // Currently we only receive the decryption key without any key backup version, it is important to
+            // check that the secret is valid for the current version before storing it.
+            // We force a check to ensure to have the latest version. We also want to check that the backup is trusted
+            // as we don't want to store the secret if the backup is not trusted, and eventually import megolm keys later from an untrusted backup.
+            const backupCheck = await this.backupManager.checkKeyBackupAndEnable(true);
+            if (backupCheck?.backupInfo?.version && backupCheck.trustInfo.trusted) {
+                const privateKeyMatcehs = this.backupManager.backupMatchesPrivateKey(backupCheck.backupInfo, value);
+
+                if (!privateKeyMatcehs) {
+                    this.logger.debug(`onReceiveSecret: backup decryption key does not match current backup version`);
+                    // just ignore the secret
+                    return;
+                }
+                this.logger.info(`onReceiveSecret: Received matching secret ${name}, store it.`);
+
+                await this.storeSessionBackupPrivateKey(decodeBase64(value), backupCheck.backupInfo.version);
+                // XXXX at this point we should probably try to download the backup and import the keys,
+                // or at least retry for the current decryption failures?
+                // Maybe add some signaling when a new secret is received, and let clients handle it?
+                // as it's where the restore from backup APIs are
+
+                // The secret is valid and stored, clear the inbox.
+                // Important to call this after storing the secret as good hygiene.
+                await this.olmMachine.deleteSecretsFromInbox("m.megolm_backup.v1");
```

Comment:
The API is a bit strange, and there is no method to just delete this secret it will delete all entries for that name, so potentially the valid secret that could be in queue.

Reply:
hrm. So if we get a bad secret, it might never be deleted?

Reply:
Yes, I am not sure how to fix that. The annoying part is that the secret handling triggers a check backup and enable.
I guess a solution could be to just listen to the secret stream to be aware of udpates, but then manually call getSecretsFromInbox that returns an array with all secrets, then delete at the end? WDYT?
The rust API is a bit strange :/